Add infrastructure proposal to discuss in-proc and out-of-proc support HLSL compiler library (#278)
microsoft/hlsl-specs: proposals/infra/INF-0004-inproc-outofproc-compiler-library-support.md

<!-- {% raw %} -->

# Support for in-process, Out-of-process, or both for the c style compiler api

* Proposal: [NNNN](INF-NNNN-inproc-outofproc-compiler-api-support.md)
* Author(s): [Cooper Partin](https://github.com/coopp)
* Sponsor: [Cooper Partin](https://github.com/coopp)
* Status: **Under Consideration**
* Impacted Project(s): (Clang)

* Issues:

## Introduction

> An effort is underway to bring compilation support into clang for HLSL based

Reviewer: We should explain here _why_ we're building this API. Why would an app prefer to use the API rather than spawning a process?

Author: ok.. I can do that.

shaders.  A C style api will be built to enable applications to compile
a shader from their own processes in addition to being able to launching the
clang process.

This document is to propose building only the out-of-process support which
gives the caller a more performant and secure solution.

Visual Studio uses an out of process architecture for compilation which helps
better utilize the power of the machine performing the compilation as well as

> isolating the code editor from any crashes that may occur during compilation.

Reviewer: I don't quite get what Visual Studio has to do with anything here?

Author: I was using it as an example.  It compiling things using out of process compilation.  I could remove this if you think it is too confusion.

Reviewer: It adds confusion.  Pretty much _anything_ that compiles by launching the compiler as a separate .exe is doing out of proc compilation.

Author: ha!  I didn't think about that!  Good point!

Author: fixed

### Overview of api hosting differences
Here are two common ways a library can be loaded and used.

* **In Process** - An application calls an api that loads compilation support
into the application's process. All work is performed within the application's
process.

* **Out of Process** - An application calls an api that launches one or more
processes to perform the compilation. All work is performed outside the
calling application's process.

## Motivation

> ### Why prefer an out of process solution?

Reviewer: Also:
* cleanup between compiler invocations
* handling of catastrophic errors (if compiler aborts on error then what does that do to the calling process?)

Author: Fixed

* Process creation on Windows based systems is expensive.  Launching clang.exe
over and over again to compile shaders becomes a lengthy operation on 
Windows platforms.

* Security is a concern when compiling a shader in a process space that is
considered protected.  An out of process design would move that concern to an
isolated process.

* In-process support using multiple threads have known thread-safety concerns
for the clang compiler.  The clang compiler holds some state at compile time
and some of that state could leak to other compilation sessions if it is not
protected. This issue was also called out when an experiment called
[llvm-buildozer](https://reviews.llvm.org/D86351) was created.  The author was
interested in multiple threads re-entering the main compiler entry point in an
effort to reduce compilation times. The effort showed that the unsafe bits
could be fixed up and made thread safe allowing the experiment to be built.
I have not checked to see the thread safe fixed bits are in the current llvm
source.

> ## Proposed solution

Reviewer: I think many of the details in here would belong in a detailed design section. I'm going to try and avoid reviewing the details here based on the experience with https://github.com/microsoft/hlsl-specs/pull/230.

### Error handling

If the HLSL compiler encounters an error during compilation or the compiler
process crashes, the rest of the compiler processes will continue on.
Error information is communicated back over the IPC mechanism to the caller
and the application will choose how to handle it.

### What if in-process becomes a requirement at a later time?

If an in process solution becomes a requirement, it can easily be built
because the support code that performs compilation will have already been
refactored out to be called by the out of process design.

### Open Questions

* Should the portions of the out of process architecture be built as a general
purpose system for others in the llvm project to be able to use in their own
systems?

## Detailed design

A detailed design of the out-of-process system has not been created.
There is an interesting example of an approach for a general purpose system
called 'Compiler Services'. This was presented as an option to support editors
and work with compiler databases. The proposal was called
[Clang C++ Services](https://github.com/chandlerc/llvm-designs/blob/master/ClangService.rst)

> ## Resources

Reviewer: The template has an "Alternatives Considered" option, which might be a good place to discuss why inproc-only or inproc+outproc were not chosen as the solution.

Author: ok.. Will do.

Author: Fixed

and compile their shaders that way?  They can and they will. Launching clang
to compile a shader will be supported.

A design requirement that cannot be satisfied with just executing clang with a 
command-line is _include handlers_. An include handler is a way for clients to
hook into the compilation process and dynamically load dependencies when
needed. We have already been asked by a game studio to support this feature.
This can only be achieved if there is an api to support adding these handlers.

### Overview of api hosting differences

Here are two common ways an api can be loaded and used.

* **In Process** - An application calls an api that loads compilation support
into the application's process. All work is performed within the application's
process.

* **Out of Process** - An application calls an api that launches one or more
processes to perform the compilation. All work is performed outside the
calling application's process.

## Motivation

### Why prefer an out of process solution?

* Process creation on Windows based systems is expensive.  Launching clang.exe
over and over again to compile shaders becomes a lengthy operation on 

> Windows platforms.

Reviewer: > Launching clang.exe over and over again to compile shaders....

This is actually describing one possible out-of-proc solution.  In-proc would not involve launching new processes.

Author: I removed this.  You are correct that it is referring to a solution and that isn't want the list is supposed to communicate.

* Security is a concern when compiling a shader in a process space that is
considered protected.  An out of process design would move that concern to an
isolated process.

> * In-process support using multiple threads have known thread-safety concerns
> for the clang compiler.  The clang compiler holds some state at compile time

Reviewer: ```suggestion
* Running multiple instances of clang In-process across multiple threads has known thread-safety concerns
for the clang compiler.  The clang compiler holds some state at compile time
```

Author: fixed

<!-- {% raw %} -->

> # In-process only or In-process and Out-of-process support for the compiler library

Reviewer: "library" is quite overloaded in this space.  Wonder if "API" would more directly indicate what this is trying to solve.

Author: Fixed up terminology usage.

> * Proposal: [0004](INF-0004-inproc-outofproc-compiler-library-support.md)

Reviewer: This proposal should start off numbered NNNN.

Reviewer: (suggest leaving any renaming of this until _just before_ this PR is completed so that it doesn't confuse the diffs)

Author: Updated the name in the document contents.  Will update the file name when the review is close to being complete to avoid diff issues.

* Author(s): [Cooper Partin](https://github.com/coopp)
* Sponsor: [Cooper Partin](https://github.com/coopp)
* Status: **Under Consideration**
* Impacted Project(s): (Clang)

* Issues:

## Introduction

An effort is underway to bring compilation support into clang for HLSL based
shaders.  A C style library will be built to enable applications to compile
a shader from their own processes in addition to being able to launching the
clang process.

There are two proposed ways one could consume the shader compiler library and

> compile a shader today.

Reviewer: > There are two proposed ways one could consume the shader compiler library and compile a shader today.

Who is the "we" in the sentence?

Is "shader compiler library" the "C style library" mentioned previously?

Author: Updated the document to be more consistent in the terms used.

* **In Process** - An application links against the shader compiler library to
bring in support for shader compilation as a static library

> (full implementation) or a dynamic library. Compilation is performed within the

Reviewer: In my mind there's a huge difference between static and dynamic here. I think that for the API we're thinking of here static linking is not on the table.

calling application's process.

> The DirectX Shader library does not currently support being linked as a

Reviewer: > The DirectX Shader library does not currently support being linked as a

What does "the DirectX Shader library" refer to?  Is it the same as the "shader compiler library" or the "C style library" mentioned earlier?  Or is this referring to DXC's API?

Author: Removed confusion.

static library.  It is a COM interface library that brings in compilation
support dynamically. Both static and dynamic in-proc solutions will be 

> supported in the HLSL clang based library.

Reviewer: > Both static and dynamic in-proc solutions will be supported in the HLSL clang based library.

Proposing that the new API will support static and dynamic is a pretty big decision being made here and so shouldn't be tucked away in the introduction.

Author: Made the proposal actually 'propose' a solution rather than be wishy washy about all possible options.

* **Out of Process** - An application links against the shader compiler library and
uses the apis to compile which create one or more processes on the
application's behalf to compile one or more shaders. Compilation is performed
outside the calling application's process and isolated away to a worker
process.

In-Process must be supported because it is already a method that applications

> use today to consume the DirectX Shader Compiler library.

Reviewer: > In-Process must be supported because it is already a method that applications use today to consume the DirectX Shader Compiler library.

I disagree with this logic.  Just because it is supported by DXC does not mean that this is the right solution for clang.

Maybe we'll still decide that we need to support in-proc, but if we do I think we need something stronger than "this is what DXC did" as a reason to do it.

Author: I removed the leaning on DXC's past as evidence to support a direction.  It should be top of mind for features, but for architecture we can be more creative as to what we want the new clang based architecture to be.

Out of process is considered an optional feature as the library could ship
without it in the first release.

This document is a discussion about an out of process design and to evaluate if

> we need to build the system immediately.

Reviewer: I think that this document needs to pick one of these possibilities:

* in-proc only
* out-of-proc only
* both in- and out-of-proc options provided

Maybe some more detail on what is meant by in- or out-of-proc would help us understand the decision.

FWIW, going in to this, "out-of-proc only" is my preferred option, so this document shouldn't assume that in-proc is required.

Author: Picked!
